### PR TITLE
Feat/custom pk flutter impl

### PR DESCRIPTION
**Pull request by HuiSF**
#### Issue #, if available

https://github.com/aws-amplify/amplify-flutter/issues/1426

#### Description of how you validated changes

<details>
<summary>Sample of generate model with a custom PK</summary>
Schema

```graphql
type Inventory @model {
  productId: ID! @primaryKey(sortKeyFields: ["productName", "warehouseId", "region"])
  productName: String!
  warehouseId: ID! @index(name: "byWarehouse", sortKeyFields: ["region"])
  region: String!
}
```

Generated Dart model

```dart
/*
* Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
*
* Licensed under the Apache License, Version 2.0 (the "License").
* You may not use this file except in compliance with the License.
* A copy of the License is located at
*
*  http://aws.amazon.com/apache2.0
*
* or in the "license" file accompanying this file. This file is distributed
* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
* express or implied. See the License for the specific language governing
* permissions and limitations under the License.
*/

// NOTE: This file is generated and may not follow lint rules defined in your app
// Generated files can be excluded from analysis in analysis_options.yaml
// For more info, see: https://dart.dev/guides/language/analysis-options#excluding-code-from-analysis

// ignore_for_file: public_member_api_docs, file_names, unnecessary_new, prefer_if_null_operators, prefer_const_constructors, slash_for_doc_comments, annotate_overrides, non_constant_identifier_names, unnecessary_string_interpolations, prefer_adjacent_string_concatenation, unnecessary_const, dead_code

import 'package:amplify_datastore_plugin_interface/amplify_datastore_plugin_interface.dart';
import 'package:flutter/foundation.dart';

/// This is an auto generated class representing the Inventory type in your schema.
@immutable
class Inventory extends Model {
  static const classType = const _InventoryModelType();
  final String? _productId;
  final String? _productName;
  final String? _warehouseId;
  final String? _region;
  final TemporalDateTime? _createdAt;
  final TemporalDateTime? _updatedAt;

  @override
  getInstanceType() => classType;

  @Deprecated(
      '[getId] is being deprecated in favor of custom primary key feature. Use getter [modelIdentifier] to get model identifier.')
  @override
  String getId() => modelIdentifier.serializeAsString();

  InventoryModelIdentifier get modelIdentifier {
    try {
      return InventoryModelIdentifier(
          productId: _productId!,
          productName: _productName!,
          warehouseId: _warehouseId!,
          region: _region!);
    } catch (e) {
      throw new DataStoreException(
          DataStoreExceptionMessages
              .codeGenRequiredFieldForceCastExceptionMessage,
          recoverySuggestion: DataStoreExceptionMessages
              .codeGenRequiredFieldForceCastRecoverySuggestion,
          underlyingException: e.toString());
    }
  }

  String get productId {
    try {
      return _productId!;
    } catch (e) {
      throw new DataStoreException(
          DataStoreExceptionMessages
              .codeGenRequiredFieldForceCastExceptionMessage,
          recoverySuggestion: DataStoreExceptionMessages
              .codeGenRequiredFieldForceCastRecoverySuggestion,
          underlyingException: e.toString());
    }
  }

  String get productName {
    try {
      return _productName!;
    } catch (e) {
      throw new DataStoreException(
          DataStoreExceptionMessages
              .codeGenRequiredFieldForceCastExceptionMessage,
          recoverySuggestion: DataStoreExceptionMessages
              .codeGenRequiredFieldForceCastRecoverySuggestion,
          underlyingException: e.toString());
    }
  }

  String get warehouseId {
    try {
      return _warehouseId!;
    } catch (e) {
      throw new DataStoreException(
          DataStoreExceptionMessages
              .codeGenRequiredFieldForceCastExceptionMessage,
          recoverySuggestion: DataStoreExceptionMessages
              .codeGenRequiredFieldForceCastRecoverySuggestion,
          underlyingException: e.toString());
    }
  }

  String get region {
    try {
      return _region!;
    } catch (e) {
      throw new DataStoreException(
          DataStoreExceptionMessages
              .codeGenRequiredFieldForceCastExceptionMessage,
          recoverySuggestion: DataStoreExceptionMessages
              .codeGenRequiredFieldForceCastRecoverySuggestion,
          underlyingException: e.toString());
    }
  }

  TemporalDateTime? get createdAt {
    return _createdAt;
  }

  TemporalDateTime? get updatedAt {
    return _updatedAt;
  }

  const Inventory._internal(
      {required productId,
      required productName,
      required warehouseId,
      required region,
      createdAt,
      updatedAt})
      : _productId = productId,
        _productName = productName,
        _warehouseId = warehouseId,
        _region = region,
        _createdAt = createdAt,
        _updatedAt = updatedAt;

  factory Inventory(
      {required String productId,
      required String productName,
      required String warehouseId,
      required String region}) {
    return Inventory._internal(
        productId: productId,
        productName: productName,
        warehouseId: warehouseId,
        region: region);
  }

  bool equals(Object other) {
    return this == other;
  }

  @override
  bool operator ==(Object other) {
    if (identical(other, this)) return true;
    return other is Inventory &&
        _productId == other._productId &&
        _productName == other._productName &&
        _warehouseId == other._warehouseId &&
        _region == other._region;
  }

  @override
  int get hashCode => toString().hashCode;

  @override
  String toString() {
    var buffer = new StringBuffer();

    buffer.write("Inventory {");
    buffer.write("productId=" + "$_productId" + ", ");
    buffer.write("productName=" + "$_productName" + ", ");
    buffer.write("warehouseId=" + "$_warehouseId" + ", ");
    buffer.write("region=" + "$_region" + ", ");
    buffer.write("createdAt=" +
        (_createdAt != null ? _createdAt!.format() : "null") +
        ", ");
    buffer.write(
        "updatedAt=" + (_updatedAt != null ? _updatedAt!.format() : "null"));
    buffer.write("}");

    return buffer.toString();
  }

  Inventory copyWith() {
    return Inventory._internal(
        productId: productId,
        productName: productName,
        warehouseId: warehouseId,
        region: region);
  }

  Inventory.fromJson(Map<String, dynamic> json)
      : _productId = json['productId'],
        _productName = json['productName'],
        _warehouseId = json['warehouseId'],
        _region = json['region'],
        _createdAt = json['createdAt'] != null
            ? TemporalDateTime.fromString(json['createdAt'])
            : null,
        _updatedAt = json['updatedAt'] != null
            ? TemporalDateTime.fromString(json['updatedAt'])
            : null;

  Map<String, dynamic> toJson() => {
        'productId': _productId,
        'productName': _productName,
        'warehouseId': _warehouseId,
        'region': _region,
        'createdAt': _createdAt?.format(),
        'updatedAt': _updatedAt?.format()
      };

  static final QueryField PRODUCTID = QueryField(fieldName: "productId");
  static final QueryField PRODUCTNAME = QueryField(fieldName: "productName");
  static final QueryField WAREHOUSEID = QueryField(fieldName: "warehouseId");
  static final QueryField REGION = QueryField(fieldName: "region");
  static var schema =
      Model.defineSchema(define: (ModelSchemaDefinition modelSchemaDefinition) {
    modelSchemaDefinition.name = "Inventory";
    modelSchemaDefinition.pluralName = "Inventories";

    modelSchemaDefinition.indexes = [
      ModelIndex(
          fields: const ["productId", "productName", "warehouseId", "region"],
          name: null),
      ModelIndex(fields: const ["warehouseId", "region"], name: "byWarehouse")
    ];

    modelSchemaDefinition.addField(ModelFieldDefinition.field(
        key: Inventory.PRODUCTID,
        isRequired: true,
        ofType: ModelFieldType(ModelFieldTypeEnum.string)));

    modelSchemaDefinition.addField(ModelFieldDefinition.field(
        key: Inventory.PRODUCTNAME,
        isRequired: true,
        ofType: ModelFieldType(ModelFieldTypeEnum.string)));

    modelSchemaDefinition.addField(ModelFieldDefinition.field(
        key: Inventory.WAREHOUSEID,
        isRequired: true,
        ofType: ModelFieldType(ModelFieldTypeEnum.string)));

    modelSchemaDefinition.addField(ModelFieldDefinition.field(
        key: Inventory.REGION,
        isRequired: true,
        ofType: ModelFieldType(ModelFieldTypeEnum.string)));

    modelSchemaDefinition.addField(ModelFieldDefinition.nonQueryField(
        fieldName: 'createdAt',
        isRequired: false,
        isReadOnly: true,
        ofType: ModelFieldType(ModelFieldTypeEnum.dateTime)));

    modelSchemaDefinition.addField(ModelFieldDefinition.nonQueryField(
        fieldName: 'updatedAt',
        isRequired: false,
        isReadOnly: true,
        ofType: ModelFieldType(ModelFieldTypeEnum.dateTime)));
  });
}

class _InventoryModelType extends ModelType<Inventory> {
  const _InventoryModelType();

  @override
  Inventory fromJson(Map<String, dynamic> jsonData) {
    return Inventory.fromJson(jsonData);
  }
}

/// This is an auto generated class representing the model identifier
/// of [Inventory] in your schema.
class InventoryModelIdentifier implements ModelIdentifier<Inventory> {
  final String productId;
  final String productName;
  final String warehouseId;
  final String region;

  /// Create an instance of InventoryModelIdentifier using [productId] the primary key.
  /// And [productName], [warehouseId], [region] the sort keys.
  const InventoryModelIdentifier(
      {required this.productId,
      required this.productName,
      required this.warehouseId,
      required this.region});

  Map<String, dynamic> serializeAsMap() {
    return {
      'productId': productId,
      'productName': productName,
      'warehouseId': warehouseId,
      'region': region
    };
  }

  List<Map<String, dynamic>> serializeAsList() {
    return [
      {'productId': productId},
      {'productName': productName},
      {'warehouseId': warehouseId},
      {'region': region}
    ];
  }

  String serializeAsString() => '$productId#$productName#$warehouseId#$region';

  @override
  String toString() =>
      'InventoryModelIdentifier(productId: $productId, productName: $productName, warehouseId: $warehouseId, region: $region)';

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }

    return other is InventoryModelIdentifier &&
        productId == other.productId &&
        productName == other.productName &&
        warehouseId == other.warehouseId &&
        region == other.region;
  }

  @override
  int get hashCode =>
      productId.hashCode ^
      productName.hashCode ^
      warehouseId.hashCode ^
      region.hashCode;
}

```

</details>

#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-codegen/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] Breaking changes to existing customers are released behind a feature flag or major version update
- [ ] Changes are tested using sample applications for all relevant platforms (iOS/android/flutter/Javascript) that use the feature added/modified

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts [hunk 1]
```diff
@@ -999,6 +1146,18 @@ export class AppSyncModelDartVisitor<
     });
   }
 
+  protected getModelIdentifierFields(model: CodeGenModel): CodeGenField[] {
```
Comment (AaronZyLee) [inline packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts:1149]: nit: This method can work until the v1 transformer is deprecated(as `@key` is v1). In long term, this method can take advantage of the change in base visitor, where the model fields can be traversed to find pk field and corresponding pk type info (including pk type and sort key fields). The pk name can also be found as well.
```TypeScript
const primaryKeyField = model.fields.find(field => field.primaryKeyInfo)!;
const { sortKeyFields } = primaryKeyField.primaryKeyInfo!
```

**Review by AaronZyLee — COMMENTED** [review]
The custom PK change LGTM. I have some questions as follows:
1. I notice some general changes such as changing the comment format to `///`, removing the `new` keyword and adding `const` keyword. It is better to include the description for these if they are intended.
2. For scenario where `id` is PK and sort key fields are defined, I can also see the sort keys are defined in the new model identifier class. However, they are not actually treated as a composite key type. I wonder what is the use case of it comparing to custom PK with SK.
3. For the getting primary key info, see the comments below.

**Comment by HuiSF** [comment]
Thanks @AaronZyLee for reviewing -

> 1. I notice some general changes such as changing the comment format to `///`, removing the `new` keyword and adding `const` keyword. It is better to include the description for these if they are intended.

Yes is to conform Dart conventions. I will separate the change out as a separate PR.

> 2. For scenario where `id` is PK and sort key fields are defined, I can also see the sort keys are defined in the new model identifier class. However, they are not actually treated as a composite key type. I wonder what is the use case of it comparing to custom PK with SK.

I wasn't aware of this scenario "where `id` is PK sort keys are not treated as a composite key type".
From what I understood, as long as a primary key (regardless the primary key filed is `id` or a custom field) and corresponding sort keys are defined in a model, the model always uses composite key as identifier.

> 3. For the getting primary key info, see the comments below.

Will update the impl.

**Review by AaronZyLee — CHANGES_REQUESTED** [review]
Could you fix the bot alert? Otherwise LGTM

**Review by AaronZyLee — APPROVED** [review]

**Pull request merged**